spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRuntimeFiltering.java

```diff
@@ -0,0 +1,386 @@
+
+    Dataset<Row> df = spark.range(1, 100)
+        .withColumnRenamed("id", "i`d")
+        .withColumn("date", date_add(expr("DATE '1970-01-01'"), expr("CAST(`i``d` % 4 AS INT)")))
+        .withColumn("ts", expr("TO_TIMESTAMP(date)"))
+        .withColumn("data", expr("CAST(date AS STRING)"))
+        .select("`i``d`", "data", "date", "ts");
+
+    df.coalesce(1).writeTo(tableName).option(SparkWriteOptions.FANOUT_ENABLED, "true").append();
+
+    sql("CREATE TABLE dim (id BIGINT, date DATE) USING parquet");
+    Dataset<Row> dimDF = spark.range(1, 2)
+        .withColumn("date", expr("DATE '1970-01-02'"))
+        .select("id", "date");
+    dimDF.coalesce(1).write().mode("append").insertInto("dim");
+
+    String query = String.format(
+        "SELECT f.* FROM %s f JOIN dim d ON f.`i``d` = d.id AND d.date = DATE '1970-01-02' ORDER BY date",
+        tableName);
+
+    assertQueryContainsRuntimeFilter(query);
+
+    deleteNotMatchingFiles(Expressions.equal("i`d", 1), 7);
+
+    assertEquals("Should have expected rows",
+        sql("SELECT * FROM %s WHERE `i``d` = 1 ORDER BY date", tableName),
+        sql(query));
+  }
+
+  @Test
+  public void testUnpartitionedTable() throws NoSuchTableException {
+    sql("CREATE TABLE %s (id BIGINT, data STRING, date DATE, ts TIMESTAMP) USING iceberg", tableName);
+
+    Dataset<Row> df = spark.range(1, 100)
+        .withColumn("date", date_add(expr("DATE '1970-01-01'"), expr("CAST(id % 4 AS INT)")))
+        .withColumn("ts", expr("TO_TIMESTAMP(date)"))
+        .withColumn("data", expr("CAST(date AS STRING)"))
+        .select("id", "data", "date", "ts");
+
+    df.coalesce(1).writeTo(tableName).append();
+
+    sql("CREATE TABLE dim (id BIGINT, date DATE) USING parquet");
+    Dataset<Row> dimDF = spark.range(1, 2)
+        .withColumn("date", expr("DATE '1970-01-02'"))
+        .select("id", "date");
+    dimDF.coalesce(1).write().mode("append").insertInto("dim");
+
+    String query = String.format(
+        "SELECT f.* FROM %s f JOIN dim d ON f.id = d.id AND d.date = DATE '1970-01-02' ORDER BY date",
+        tableName);
+
+    assertQueryContainsNoRuntimeFilter(query);
+
+    assertEquals("Should have expected rows",
+        sql("SELECT * FROM %s WHERE id = 1 ORDER BY date", tableName),
+        sql(query));
+  }
+
+  private void assertQueryContainsRuntimeFilter(String query) {
+    assertQueryContainsRuntimeFilters(query, 1);
+  }
+
+  private void assertQueryContainsNoRuntimeFilter(String query) {
+    assertQueryContainsRuntimeFilters(query, 0);
+  }
+
+  private void assertQueryContainsRuntimeFilters(String query, int expectedFilterCount) {
+    List<Row> output = spark.sql("EXPLAIN EXTENDED " + query).collectAsList();
+    String plan = output.get(0).getString(0);
+    int actualFilterCount = StringUtils.countMatches(plan, "dynamicpruningexpression");
+    Assert.assertEquals("Invalid number of runtime filters", expectedFilterCount, actualFilterCount);
```

Comment:
Minor: this isn't a great error message for debugging, since we know that there should either be a filter or not filters. I'd probably pass the message in from `assertQueryContainsRuntimeFIlter` so you can customize: "Query should have one runtime filter"

Reply:
Fixed.